### PR TITLE
Th2 3007

### PR DESCRIPTION
**Pull request by Nikita-Smirnov-Exactpro**

### CODE REVIEW

File: src/test/kotlin/com/exactpro/th2/codec/TestSyncDecoder.kt [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+        val captor = argumentCaptor<MessageGroupBatch>()
+        verify(router).sendAll(captor.capture(), any())
+
+        val result = captor.lastValue
+
+        Assertions.assertEquals(1, result.groupsCount) { "Groups count: ${shortDebugString(result)}" }
+        val group = result.getGroups(0)
+        Assertions.assertEquals(3, group.messagesCount) { "Messages count: ${shortDebugString(group)}" }
+
+        val first = group.getMessages(0)
```
Comment (Topru333) [inline src/test/kotlin/com/exactpro/th2/codec/TestSyncDecoder.kt:89]:
```suggestion
        group.getMessages(0).let { message -> ...test }
```

Comment (Topru333) [reply src/test/kotlin/com/exactpro/th2/codec/TestSyncDecoder.kt:89]: same can be apply for first second third to remove vals

File: src/main/kotlin/com/exactpro/th2/codec/SyncDecoder.kt [hunk 2]
```diff
@@ -42,19 +40,25 @@ class SyncDecoder(
...
             }
+            groupBuilder.addMessages(notTypeMessage)
         }
 
         return if (groupBuilder.messagesCount > 0) groupBuilder.build() else null
     }
 
+    private fun checkProtocol(rawMessage: RawMessage) = rawMessage.metadata.protocol.let {
+        it.isNullOrEmpty() || it == processor.protocol
+    }
```
Comment (ivandruzhinin) [inline src/main/kotlin/com/exactpro/th2/codec/SyncDecoder.kt:60]: Why do we even check protocol of raw messages? Who's gonna set it? In `th2-codec` we don't check protocol of raw messages during decode at all

File: src/main/kotlin/com/exactpro/th2/codec/SyncEncoder.kt [hunk 3]
```diff
@@ -42,13 +42,19 @@ class SyncEncoder(
...
             }
+            groupBuilder.addMessages(notTypeMessage)
         }
 
         return groupBuilder.build()
     }
+
+    private fun checkProtocol(message: Message) = message.metadata.protocol.let {
+        it.isNullOrEmpty() || it == processor.protocol
+    }
```
Comment (ivandruzhinin) [inline src/main/kotlin/com/exactpro/th2/codec/SyncEncoder.kt:58]: I think we should check protocol the same way we do in `th2-codec`: https://github.com/th2-net/th2-codec/blob/master/src/main/kotlin/com/exactpro/th2/codec/EncodeProcessor.kt#L72

File: src/main/kotlin/com/exactpro/th2/codec/SyncDecoder.kt (hunk 2 above)
Comment (Nikita-Smirnov-Exactpro) [reply src/main/kotlin/com/exactpro/th2/codec/SyncDecoder.kt:60]: In some cases, th2-read can produce messages of several protocols in one th2 stream, for example, FIX and TCP, where TCP contains information about connection lifecycle.

Comment (Nikita-Smirnov-Exactpro) [reply src/main/kotlin/com/exactpro/th2/codec/SyncDecoder.kt:60]: Sorry, you right messages with different protocols should be placed in different groups

**Review by Topru333 — APPROVED** [review]

**Review by ivandruzhinin — APPROVED** [review]

**Pull request merged**